### PR TITLE
Consolidate code into modules

### PR DESCRIPTION
**Pull request by travismyers19**
Reorganize code, remove path hacking, and add command line arguments

- Move all scripts to top directory of project directory

- Modify all scripts that import modules to import from Modules

- Modify all scripts to take command line arguments

- Create "Addresses" folder to store csv files of addresses

- Create empty "Data" folder to be used for storing collected data

- Add "requirements.txt" to ease installation in non-conda environments

- Update "README.md"

--------------------------------------------

Remove machine-specific paths, clean up code, and finish readme file

- build/environment.sh: replace machine-specific path with $(pwd)

- Small Data: add a small amount of data so the user can test to see if the repo is working properly

- Training/* : remove machine-specific paths

- README.md: finish writing

-------------------------------------

- Add buildingclassifier.py which contains the class BuildingClassifier which provides functionality for creating and training the model

- Add customlosses.py which contains loss functions for binary and categorical crossentropy modified for focal loss in order to
combat class imbalance

- Add imagefunctions.py which contains functions for reading and writing a single image

- Expand Training/train.sh to include more configurations

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,73 @@
...
+
+##Installation and Dependencies
+Clone the repo
+
+##Preparing the Environment
+Before running any scripts in this repo, source the "build/environment.sh" bash script to add "Modules" to the PYTHONPATH environment variable:
+
+```
+source build/environment.sh
+```
```
Comment (AntonMu) [inline README.md]: Ideally, you don't want to do this kind of "sys path hacking". It makes your code very fragile to small changes. For instance, in your script you assume that the user has a specific folder name (`$HOME/Insight/Vulnerable_Buildings`) - you shouldn't assume this. 

Also, this path hacking makes your code less modular - if another code would like to use that code and import it, it wouldn't work. 

Here is a better way: 
 * Put your main (highest level) executables (train, preprocess etc) in the top level of your directory and then use module imports in your file, i.e.
`from Modules import addresses` 

This is not a high priority - but keep it in mind in the future or if there is time left.

File: README.md [hunk 2]
```diff
@@ -1,5 +1,73 @@
...
+
+##Modules
+The "Modules" folder contains all of the custom modules in this repo.
+
+###addresses.py
```
Comment (AntonMu) [inline README.md]: Use
```
### `addresses.py`
```

File: README.md [hunk 3]
```diff
@@ -1,5 +1,73 @@
...
+This module contains the class Addresses which provides functionality for grabbing random addresses using reverse geocoding and for getting images from Google Street View corresponding to given addresses.
+
+###buildingclassifier.py
+This module contains the class BuildingClassifier which provides functionality for creating and training custom Inception V3 models for either ternary classification or binary classification, as well as functions for evaluating a model to determine the following statistics given a directory containing test images:
+
+Accuracy:  Percentage of the test images labeled correctly.
+Soft Precision:  Precision in determining soft vs. non-soft.
+Soft Recall:  Recall in determining soft vs. non-soft.
+Good Precision:  Precision in determining good image vs. bad image.
+Good Recall:  Recall in determining good image vs. bad image.
```
Comment (AntonMu) [inline README.md]: Make it a list (see above)

File: Training/train.sh [hunk 4]
```diff
@@ -2,24 +2,41 @@
...
 #The location of the data for training.
-export DATA_FOLDER="/home/ubuntu/Insight/s3mnt/data2/preprocessed/"
+export TRAINING_DIRECTORY="/home/ubuntu/Insight/s3mnt/Training_Soft/"
+
+#The location of the data for testing/validation.
+export TEST_DIRECTORY="/home/ubuntu/Insight/s3mnt/Test_Soft_Binary/"
 
 #The location of the model to train
-export MODEL="/home/ubuntu/Insight/Vulnerable_Buildings/Models/inception_model.h5"
+export MODEL_FILENAME="/home/ubuntu/Insight/s3mnt/Models/BinaryFocalLoss/soft_model.h5"
```
Comment (AntonMu) [inline Training/train.sh]: Again, these paths are too specific for your case. 

Replace these statements with something that relates to your current working directory. For example

```
export PYTHONPATH=$PYTHONPATH:`pwd`
```
Get's the current directory in bash style. 

Generally, using bash scripts this way is not a good way. Try to put it into a python file instead `train.py` and then use command line arguments with good defaults to add (optional) parameters. For example, this is one of my [`train.py`](https://github.com/AntonMu/TrainYourOwnYOLO/blob/master/2_Training/Train_YOLO.py)

File: build/environment.sh [hunk 5]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+export PYTHONPATH=$HOME/Insight/Vulnerable_Buildings/Modules:$PYTHONPATH
```
Comment (AntonMu) [inline build/environment.sh]: See comment above - use something that would work for anyone using this code. 
E.g. put this at the top level of your repo. 
```
export PYTHONPATH=$PYTHONPATH:`pwd`
```
Then for imports use `import Modules` at the top of your python file. 

Ideally, don't use this path hacking at all.

File: Modules/buildingclassifier.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        model.compile(optimizer=opt, loss=loss, metrics=the_metrics)
+        model.save(self.model_filename)
+<<<<<<< HEAD
+        self.model = model
+
+    def train_model(self, epochs, batch_size, training_directory, test_directory, trained_model_filename, metrics_filename, binary=False, weights=None):
+=======
+
+    def train_model(self, epochs, batch_size, training_directory, test_directory, trained_model_filename, metrics_filename, binary=False):
+>>>>>>> 6ba2785555e39efda145a8fe1006732a213805e7
```
Comment (AntonMu) [inline Modules/buildingclassifier.py]: Looks like you have merge conflicts here.

**Comment by AntonMu** [comment]
Approved - feel free to merge!
